### ISSUE TITLE
Better handle PII being turned off

### ISSUE THREAD

**Issue by bmtcril** [openedx/tutor-contrib-aspects#961]
Right now PII is off by default, however we have 2 dashboards that rely heavily on that data and look broken without it. We have 2 choices that I can see:

1- Turn those dashboards off when PII is off (but allow overrides)
2- Replace the PII with some kind of placeholders

Complications for 2:

- In order to make the cross filters work they would need to be unique somehow per user, but not in a way that exposes PII
- Any strings like "unknown" or "hidden" that we might need to use would need to be translatable, which can be a challenge

A possible way to do this is take some part of the user's per-course UUID and use that when the PII isn't present, but that does get hard to read and unless we hash it somehow will potentially expose enough of that key to be a privacy concern.

We should try to come up with a solution for this in the short term as it's a pretty painful issue.

**Comment by felipemontoya** [comment]
Hey, I'm picking this up as epic 4 of **FC0140**.

Before proposing anything I set up a Verawood + Aspects environment, generated synthetic data (including at-risk learners in a demo course), and walked through both dashboards with PII on and off, so the notes below are from observed behaviour rather than from reading the code alone.

## Where I agree with what's already here

Sarah and Ty's conclusion — don't show the **Individual Learner** dashboard when PII is off — is right, and it should be part of what we do. Without identity the dashboard can't serve its purpose, and the Course Dashboard covers the aggregate view.

The original product framing for this work was:

> As an instance operator, I want to hide the Individual Learner Dashboard from the Instructor Dashboard tabs shown on the Reports tab of the Instructor Dashboard, so that my course teams aren't presented with Dashboards that they can't use on the Instructor Dashboard.

I think we can do better than that, for two reasons.

## Reason 1: hiding the tab doesn't hide the data

Removing the dashboard from the Reports tab changes what course teams *see there*. It doesn't change what's in ClickHouse. Anyone with Superset access — and on many installs that includes staff with a legitimate reason to be in Superset — can still query `event_sink.user_pii` directly, or open the same dashboard in Superset where it hasn't been removed.

Two related things I confirmed while testing:

- **Turning the flag off doesn't remove anything already collected.** An install that ran with `ASPECTS_ENABLE_PII: true` and later turns it off keeps every name and email already in ClickHouse. New rows arrive blank, old rows keep their identities, and the same table shows both at once with nothing explaining why.
- **The Individual Learner dashboard isn't the only PII surface.** `at_risk_learners.sql` selects `username`, `name` and `email` too, so the At-Risk dashboard stays in the Reports tab still showing identities. Fixing only the Individual Learner dashboard leaves that one behind.

## Reason 2: the flag isn't actually enforced everywhere

`ASPECTS_ENABLE_PII: False` should mean no PII reaches ClickHouse by any route. Right now `is_enabled()` is only consulted on the signal-driven path (`tasks.py`). Two other writers bypass it:

- **`dump_data_to_clickhouse`** (in `platform-plugin-aspects`) never checks whether the sink is enabled. `tutor local do dump-data-to-clickhouse --object user_profile` will happily populate the PII tables on an install where the flag is off. (`--force` is unrelated — it only re-serializes recently dumped objects.) I think this is a straightforward bug.
- **`load-xapi-test-data` / `xapi-db-load`** writes generated profiles straight into ClickHouse over the HTTP interface, bypassing the sinks entirely. Lower stakes because it's synthetic data, but it was confusing in my dev environment as it looked like PII was enabled when it never had been.

Happy to file these separately if that's preferred. I mention them here because they're part of why "the flag is off" can't be assumed to mean "there's no PII".

## What I'd propose

Keep the removal, and add a conditional rendering path so the data stays usable without exposing identities:

1. **Hide the Individual Learner dashboard from the Instructor Dashboard when PII is off**, overridable by the operator as everything already is.
2. **Make the PII-dependent queries conditional on the flag**, swapping the `username` / `name` / `email` columns for a single pseudonymous identifier when PII is disabled. This is what keeps the At-Risk dashboard (and the Superset copy of Individual Learner) usable rather than broken.
3. **Derive that identifier as a hash, not as a slice of the UUID.** Something like `sipHash64(external_user_id + course_key + <per-install secret>)`, truncated for readability. The per-install secret makes it irreversible to a Superset user; scoping it to the course means the same learner gets different identifiers in different courses, so nothing correlates across courses.
4. **Fix the filters at the same time.** `user_filters.sql` builds the Username / Name / Email filter options from `user_pii`, so with PII off those dropdowns are empty and because the fragment expands to `and actor_id in (select … from user_pii …)`, an applied value zeroes out the chart instead of being ignored. Those three filters should collapse into one identifier filter. This also fixes the cross-filter problem described in the original issue, where every learner shares the same empty username and clicking one selects all of them.
5. **Change `enrollment_pii.sql` from an inner join to a left join.** Today it inner-joins `dim_user_pii`, so with PII off it returns zero rows rather than rows with blanked columns. In this case the chart reads "No results were returned for this query" even though the enrollments exist. Worth noting the two join styles currently fail differently, which makes the behaviour hard to explain to operators.


## Note about the hash

What I propose would also be pseudonymisation, not anonymisation. A Superset end user would not be able to tie the displayed identifier back to the original anonymous ID, let alone to a person, which is the main thing we're protecting against. But someone who also has Instructor Dashboard access to a small course could still correlate timestamps and activity between the two and make a reasonable guess at who a row belongs to. That's outside what Aspects can prevent on its own. The goal I'd suggest is making the Aspects side sound, and being clear about where the boundary is. If this is PII for GDPR is a questions for lawyers and I'm happily not one.


## One more thing I noticed

`at_risk_learner_filter.sql` inner-joins page visits, so a learner with **no** page views at all can never be flagged at risk — the most disengaged learner in a course is invisible to the At-Risk dashboard. Unrelated to PII; happy to open a separate issue if that's useful.

## Implementation note

When thinking about this, my own obvious objection to (2) is that the assets are Jinja-rendered by Tutor and baked into the `aspects-superset` image, so flipping the flag would mean rebuilding the image and re-importing assets. That's avoidable if the branch is evaluated at **query time** rather than render time. `superset_config_docker.py` already defines `JINJA_CONTEXT_ADDONS` (`can_view_courses`, `translate_column`, …) and the `pythonpath` directory is bind-mounted into the Superset containers, so exposing the flag as an addon and putting the branch inside a `{% raw %}` block means changing policy costs a `tutor config save` and a Superset restart — no rebuild, no `import-assets`. The instructor dashboard list works the same way, since `ASPECTS_INSTRUCTOR_DASHBOARDS` renders straight into the LMS settings.

That matters because operators often don't understand the implications of Aspects until it's been running for a while. When they do change their mind, acting on it should be cheap. We for example, on our hosting solution, normally enable PII when first asked to put aspects.


cc @crathbun428

**Comment by saraburns1** [comment]
> at_risk_learner_filter.sql inner-joins page visits, so a learner with no page views at all can never be flagged at risk — the most disengaged learner in a course is invisible to the At-Risk dashboard. Unrelated to PII; happy to open a separate issue if that's useful.

This is expected. Instructors want to be able to identify users who HAVE interacted with the course at some point and then stopped. This helps clear out the cases where someone enrolled and never opened the course (that requires different communication).

**Comment by saraburns1** [comment]
> load-xapi-test-data / xapi-db-load writes generated profiles straight into ClickHouse over the HTTP interface, bypassing the sinks entirely. Lower stakes because it's synthetic data, but it was confusing in my dev environment as it looked like PII was enabled when it never had been.

I wouldn't worry about this, it's mostly for us testing performance. As you said, all of the data is synthetic anyway

**Comment by saraburns1** [comment]
> Make the PII-dependent queries conditional on the flag, swapping the username / name / email columns for a single pseudonymous identifier when PII is disabled. This is what keeps the At-Risk dashboard (and the Superset copy of Individual Learner) usable rather than broken.

So the username/name/email would still be stored when PII is enabled, and then if it is disabled, the query would combine them into some hash for display in the dashboard? Wouldn't that still pose the issue that you brought up where someone with Superset access could still query that information?

**Comment by felipemontoya** [comment]
> So the username/name/email would still be stored when PII is enabled

No, not at all. Right now, when PII is disabled, no new PII flows from mysql->clickhouse over the event stream. Actually I want to strengthen this further by also making the `dump_data_to_clickhouse` command to also respect the ENABLE_PII flag.

>  the query would combine them into some hash for display

The query would combine `external_user_id` + `course_key` + `<per-install secret>` for rendering. This way instead of tables with emtpy username, name, email, we would get the pseudonymous ID. That way there is no PII data, but the user identifier is not emtpy (which causes a lot of havoc in the dynamic filtering of the dashboards in superset)

**Comment by felipemontoya** [comment]
I came back to this with a different thought and now I notice that maybe I didn't explain myself right.

Right now when PII_enabled pii data is copied to a table in clickhouse that holds PII + external_id. When PII_disabled, data is no longer copied.
This has the implication that when PII enabled for some time and at some point disabled, the table already contains some data. datasets, charts and dashboards will happily display it. New external_ids, that hold no PII data, appear as empty rows in the charts (at least for the PII columns).

The way I understand it, disabling PII Signals an intention from the operators to no longer share this data. We could take the stand that doing so should also truncate the external ID table so that even data that was previously sent to Clickhouse is removed. We don't do that yet, but I'm happy to create a PR for that as well.

However my proposal above is only to stop looking up the PII data and instead show the calculated pseudonymous.

**Comment by felipemontoya** [comment]
From our meeting. Approved tasks:

- hide the individual learner
- truncate external_id table (+user_profile)
- make dump_data respect PII flag
